grpc-gcp/src/test/java/com/google/cloud/grpc/ChannelIdPropagationTest.java

```diff
@@ -0,0 +1,108 @@
+    @Override
+    public InputStream stream(T value) {
+      return null;
+    }
+
+    @Override
+    public T parse(InputStream stream) {
+      return null;
+    }
+  }
+
+  @Test
+  public void testChannelIdKeySetWithoutAffinity() {
+    ManagedChannelBuilder<?> builder = ManagedChannelBuilder.forAddress("localhost", 443);
+    final AtomicInteger channelId = new AtomicInteger(-1);
+
+    ClientInterceptor channelIdInterceptor =
+        new ClientInterceptor() {
+          @Override
+          public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+              MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
+            return new ForwardingClientCall.SimpleForwardingClientCall<ReqT, RespT>(
+                next.newCall(method, callOptions)) {
+              @Override
+              public void start(Listener<RespT> responseListener, Metadata headers) {
+                // System.out.println("CALL OPTIONS(CHANNELID) IS NULL " +
+                // (callOptions.getOption(GcpManagedChannel.CHANNEL_ID_KEY) == null));
+                if (callOptions.getOption(GcpManagedChannel.CHANNEL_ID_KEY) != null) {
+                  channelId.set(callOptions.getOption(GcpManagedChannel.CHANNEL_ID_KEY));
+                }
+                super.start(responseListener, headers);
+              }
+            };
+          }
+        };
+
+    // CORRECTED: Add interceptor to the delegate builder so it runs on the underlying channels
+    builder.intercept(channelIdInterceptor);
+
+    // Creating a pool.
+    final GcpManagedChannel pool =
+        (GcpManagedChannel)
+            GcpManagedChannelBuilder.forDelegateBuilder(builder)
+                .withOptions(
+                    GcpManagedChannelOptions.newBuilder()
+                        .withChannelPoolOptions(
+                            GcpChannelPoolOptions.newBuilder().setMinSize(3).setMaxSize(3).build())
+                        .build())
+                .build();
+
+    MethodDescriptor<String, String> methodDescriptor =
+        MethodDescriptor.<String, String>newBuilder()
+            .setType(MethodDescriptor.MethodType.UNARY)
+            .setFullMethodName("test/method")
+            .setRequestMarshaller(new FakeMarshaller<>())
+            .setResponseMarshaller(new FakeMarshaller<>())
+            .build();
+
+    // Use the pool directly (interceptor is already inside)
+    ClientCall<String, String> newCall = pool.newCall(methodDescriptor, CallOptions.DEFAULT);
+    Metadata headers = new Metadata();
+
+    // First call (should initialize channel and correct ID)
+    newCall.start(
+        new ForwardingClientCall.SimpleForwardingClientCall.Listener<String>() {}, headers);
+
+    // We expect it to be set even on the first call if everything is wired correctly,
+    // but the user's test checked it after the second call. The channel ID assignment happens in
```

Comment:
the channels assigned would be 0, 1, 2. So here we can test:
`assertThat(channelId.get()).isEqualTo(0);`